// Cloud Discovery Service - Fleet-wide service registry queries
// Provides gRPC API for querying aggregated service data from PostgreSQL
syntax = "proto3";

package ifex.cloud.discovery;

import "discovery-sync-envelope.proto";

// === Vehicle Summary ===

message VehicleSummary {
    string vehicle_id = 1;
    string fleet_id = 2;
    string region = 3;
    string model = 4;
    int32 year = 5;
    int32 service_count = 6;
    int32 job_count = 7;
    uint64 last_seen_ns = 8;
    bool is_online = 9;
}

// === Service Location ===

message ServiceLocation {
    string vehicle_id = 1;
    string fleet_id = 2;
    string region = 3;
    swdv.discovery_sync_envelope.service_info_t service = 4;
}

// === Service Statistics ===

message ServiceStats {
    string service_name = 1;
    int32 vehicle_count = 2;           // How many vehicles have this service
    int32 available_count = 3;         // How many are currently available
    map<string, int32> by_version = 4; // Version distribution
    map<string, int32> by_region = 5;  // Regional distribution
}

// === List Vehicles ===

message ListVehiclesRequest {
    string fleet_id_filter = 1;        // Optional: filter by fleet
    string region_filter = 2;          // Optional: filter by region
    bool online_only = 3;              // Only vehicles seen recently
    bool with_services_only = 4;       // Only vehicles with registered services
    int32 page_size = 10;              // Default: 50
    string page_token = 11;            // Pagination token
}

message ListVehiclesResponse {
    repeated VehicleSummary vehicles = 1;
    string next_page_token = 2;
    int32 total_count = 3;
}

// === Get Vehicle Services ===

message GetVehicleServicesRequest {
    string vehicle_id = 1;
}

message GetVehicleServicesResponse {
    string vehicle_id = 1;
    string fleet_id = 2;
    string region = 3;
    repeated swdv.discovery_sync_envelope.service_info_t services = 4;
    uint64 last_sync_ns = 5;
    uint32 state_checksum = 6;
}

// === Query Services By Name ===

message QueryServicesByNameRequest {
    string service_name_pattern = 1;   // Supports wildcards (e.g., "climate*", "*_service")
    string fleet_id_filter = 2;        // Optional
    string region_filter = 3;          // Optional
    bool available_only = 4;           // Only services with status=AVAILABLE
    int32 page_size = 10;
    string page_token = 11;
}

message QueryServicesByNameResponse {
    repeated ServiceLocation locations = 1;
    string next_page_token = 2;
    int32 total_count = 3;
}

// === Get Fleet Service Stats ===

message GetFleetServiceStatsRequest {
    string fleet_id_filter = 1;        // Optional
    string region_filter = 2;          // Optional
}

message GetFleetServiceStatsResponse {
    repeated ServiceStats stats = 1;
    int32 total_vehicles = 2;
    int32 online_vehicles = 3;
    int32 total_services = 4;
}

// === Find Vehicles With Service ===

message FindVehiclesWithServiceRequest {
    string service_name = 1;           // Exact service name
    string method_name = 2;            // Optional: must have this method
    string fleet_id_filter = 3;
    string region_filter = 4;
    bool available_only = 5;
    int32 page_size = 10;
    string page_token = 11;
}

message FindVehiclesWithServiceResponse {
    repeated VehicleSummary vehicles = 1;
    string next_page_token = 2;
    int32 total_count = 3;
}

// === gRPC Service Definition ===

service CloudDiscoveryService {
    // List all vehicles with summary info
    rpc ListVehicles(ListVehiclesRequest) returns (ListVehiclesResponse);

    // Get detailed services for a specific vehicle
    rpc GetVehicleServices(GetVehicleServicesRequest) returns (GetVehicleServicesResponse);

    // Query services across fleet by name pattern
    rpc QueryServicesByName(QueryServicesByNameRequest) returns (QueryServicesByNameResponse);

    // Get fleet-wide service statistics
    rpc GetFleetServiceStats(GetFleetServiceStatsRequest) returns (GetFleetServiceStatsResponse);

    // Find all vehicles that have a specific service
    rpc FindVehiclesWithService(FindVehiclesWithServiceRequest) returns (FindVehiclesWithServiceResponse);
}
